### PR TITLE
Avro timestamp parsing and Avro to Parquet writing using Confluent Schema Registry, JSON to Parquet

### PR DESCRIPTION
**Pull request by achad4**
- Use Confluent’s CachedSchemaRegistryClient to allow Avro message parsing
- Add Avro to Parquet writing 
- Extend the current Protobuf reading and writing capabilities to allow writing Parquet files from JSON messages

### CODE REVIEW

File: src/main/java/com/pinterest/secor/common/SecorConfig.java [hunk 1]
```diff
@@ -274,7 +276,7 @@ public String getAwsSecretKey() {
     }
 
     public String getAwsSessionToken() {
-        return getString("aws.session.token");
+        return getString("aws.session.token", "");
```
Comment (HenryCaiHaiying) [inline src/main/java/com/pinterest/secor/common/SecorConfig.java]: Why this change?

File: src/main/java/com/pinterest/secor/common/SecorSchemaRegistryClient.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        } catch (Exception e){
+            LOG.error("Error initalizing schema registry", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    public GenericRecord decodeMessage(String topic, byte[] message) {
+        GenericRecord record = (GenericRecord) decoder.fromBytes(message);
+        Schema schema = record.getSchema();
+        schemas.putIfAbsent(topic, schema);
```
Comment (HenryCaiHaiying) [inline src/main/java/com/pinterest/secor/common/SecorSchemaRegistryClient.java]: Does this mean you will not update the topic schema to latest?

File: src/main/java/com/pinterest/secor/common/SecorConfig.java (hunk 1 above)
Comment (achad4) [reply src/main/java/com/pinterest/secor/common/SecorConfig.java]: I was unable to start the application when I didn't set this parameter. It shouldn't be required right?

File: src/main/java/com/pinterest/secor/common/SecorSchemaRegistryClient.java (hunk 2 above)
Comment (achad4) [reply src/main/java/com/pinterest/secor/common/SecorSchemaRegistryClient.java]: That's a good point @HenryCaiHaiying, I will add unit tests to this class for that scenario.

**Comment by achad4** [comment]
@HenryCaiHaiying thank you for reviewing! I managed to write JSON to parquet by extending the ProtobufUtil (see` Message decodeJsonMessage(String topic, byte[] payload)`). Per the documentation I added, this functionality will only kick in when explicitly set. 

Today, I actually extended some unit tests and made a couple changes which I plan to push in the morning. I'm looking forward to your feedback.

File: src/main/java/com/pinterest/secor/common/SecorSchemaRegistryClient.java (hunk 2 above)
Comment (HenryCaiHaiying) [reply src/main/java/com/pinterest/secor/common/SecorSchemaRegistryClient.java]: So you don't want to update the schema?

File: src/main/java/com/pinterest/secor/common/SecorConfig.java (hunk 1 above)
Comment (HenryCaiHaiying) [reply src/main/java/com/pinterest/secor/common/SecorConfig.java]: We usually want those parameters to be explicitly set in your config file rather than java code defaulting.  But if you are not using aws, maybe we should not require this as a required config.

File: src/main/java/com/pinterest/secor/common/SecorSchemaRegistryClient.java (hunk 2 above)
Comment (achad4) [reply src/main/java/com/pinterest/secor/common/SecorSchemaRegistryClient.java]: No, I do want to. I'll make changes and add tests to ensure they are updated. The SchemaRegistryClient should already handle changes, so this wrapper may even be unnecessary.

**Comment by HenryCaiHaiying** [comment]
For the JSON->Parquet, it's a bit odd to use Protobuf, would JSON->Avro->Parquet work (that seems what most people was doing)?  In either situations, how do you specify the schema for JSON?

File: src/main/java/com/pinterest/secor/common/SecorConfig.java (hunk 1 above)
Comment (achad4) [reply src/main/java/com/pinterest/secor/common/SecorConfig.java]: I am using AWS but not a session token. Just permanent API keys and secrets. In that scenario I don't think think should be required.

**Comment by achad4** [comment]
That's interesting, and I'd definitely be open to switching. I went with Protobuf, because Secor already had an established workflow from Protobuf to Parquet and I didn't want to add too much overhead.
You just define the protobuf schema as you normally would, and the JSON is mapped in memory to a protobuf message using this class https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/util/JsonFormat.Parser. 
The only additional configurations you need are to define which topics are using JSON (as described in the documentation I added). 
- `secor.topic.message.format.<topic>=JSON` or`secor.topic.message.format.*=JSON`

**Comment by achad4** [comment]
@HenryCaiHaiying I'll push my JSON -> Protobuf changes and tests sometime tomorrow. I actually now have my latest version of this running in production on 6 JSON topics averaging together about 10000 messages per second.

File: src/main/java/com/pinterest/secor/common/SecorSchemaRegistryClient.java (hunk 2 above)
Comment (achad4) [reply src/main/java/com/pinterest/secor/common/SecorSchemaRegistryClient.java]: @HenryCaiHaiying I fixed and tested this behavior: please see the new `SecorSchemaRegistryClientTest`

File: src/main/java/com/pinterest/secor/common/SecorConfig.java (hunk 1 above)
Comment (HenryCaiHaiying) [reply src/main/java/com/pinterest/secor/common/SecorConfig.java]: That's reasonable.

**Comment by HenryCaiHaiying** [comment]
That's for the work, this is a good contribution.  Other people have waited for Avro and JSON for quite a while now.

**Comment by achad4** [comment]
Thanks @HenryCaiHaiying! I don't think a pure avro implementation would be too tough.
